A83-xds-gcp-authn-filter.md

```diff
@@ -0,0 +1,330 @@
+instead of the expiration time encoded in the token.
+
+When the credential is asked for a token for a data
+plane RPC, if the token is not yet cached or the cached
+token will expire within some fixed refresh interval
+(typically 1 minute), the credential will start an HTTP request to
+`http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/identity?audience=[AUDIENCE]`,
+where `[AUDIENCE]` is replaced with the audience specified when the
+credential object was instantiated.  The HTTP request will include the
+header `Metadata-Flavor: Google`.
+
+When a data plane RPC starts, if the token is cached and is not expired,
+the token will immediately be added to the RPC, and the RPC will continue.
+Otherwise (i.e., before the token is initially obtained or after the
+cached token has expired), the data plane RPC will be queued until the
+HTTP request completes.  When the HTTP request completes, the result
+(either success or failure, as described below) will be applied to all
+queued data plane RPCs.
+
+Note that when the token's expiration time is less than the refresh
+interval in the future, a new data plane RPC being started will trigger
+a new HTTP request, but the cached token value will still be used for
+that data plane RPC.  This pre-emptive re-fetching is intended to avoid
+periodic latency spikes when refreshing the token.
+
+If the HTTP request fails, all queued data plane RPCs will be failed
+with a gRPC status determined based on the returned HTTP status.  If the
+returned HTTP status maps to `UNAVAILABLE` in [HTTP to gRPC Status Code
+Mapping](https://github.com/grpc/grpc/blob/master/doc/http-grpc-status-mapping.md),
+then the data plane RPCs will be failed with status `UNAVAILABLE`;
+otherwise, they will be failed with status `UNAUTHENTICATED`.  If the
+request fails without an HTTP status (e.g., an I/O error), all queued
+data plane RPCs will be failed with `UNAVAILABLE` status.
+
+If the HTTP request succeeds, the body of the response will contain the
+JWT token. which the client will cache.  The client does not need to
+do full [RFC-7519] validation of the token (that is the responsibility
+of the server side), but it does need to extract the `exp` field for
+caching purposes.  If the `exp` field cannot be extracted (i.e., the JWT
+token is invalid), all queued data plane RPCs will be failed with status
+`UNAUTHENTICATED`.  Otherwise, the cache is updated, and the returned
+token is added to all queued data plane RPCs, which may then continue.
+
+If the HTTP request does not result in the cache being updated (i.e.,
+if the HTTP request fails or if it returns an invalid JWT token),
+[backoff](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md)
+must be applied before the next attempt may be started.  If a data
+plane RPC is started when there is no cached token available and while
+in backoff delay, it will be queued until the completion of the next
```

Comment:
Hmm.  That's a good point, and I think I agree that we should ideally fail RPCs in this case.

The reason we did it this way was so that we have some notion of when to reset the backoff state.  When the backoff expires, if there are queued RPCs, then we immediately trigger another HTTP fetch, but if there are *no* queued RPCs, then we clear the backoff state.  If we didn't do that, then there could be situations where we have a couple of failures and get the backoff delay up, but then a day goes by without another RPC, and then the next RPC that sees a failure causes us to reuse the day-old backoff state.

But thinking about this now, I think there are better alternatives here.  The obvious one is that we could just accept that we're using day-old backoff state in that scenario.  If the server has actually been down the whole time, then that's actually not inappropriate.  And even if it came back up while we weren't trying but has now gone down again, that's a sufficiently coincidental case that it's probably fine to treat it as if it had been down the whole time.

Or, if we really want to avoid this, we could have some fixed timeout for resetting the backoff state if we haven't made another attempt in a long enough period of time.  I'm thinking maybe something like 30 minutes.

Thoughts...?

Reply:
> but then a day goes by without another RPC, and then the next RPC that sees a failure causes us to reuse the day-old backoff state.

Not actually? That RPC will cause us to fetch a token, starting immediately. If that succeeds, then the overly large backoff is never used. If it fails, then using a large backoff is still fine.

(Compared to reconnect handling, this will have blips of increased RPC latency as (fail to) fetch the token. But that's acceptable and core to the approach taken here.)

Reply:
Right, sorry, I meant the RPC *after* the one that occurs after a day of being idle.  The one that occurs after a day of being idle will definitely be dispatched immediately, but if we retain backup state from the previous day, then the delay after that failure will be longer than maybe it should be.

But it sounds like we're converging on that being okay.  I've updated the doc accordingly, and I will also update the C-core implementation.